### PR TITLE
feat: Unique encrypted index+value pair support

### LINKED ISSUE #57: Unique encrypted index+value pair support
The EDV specification allows for an encrypted index+value pair to be declared as unique. Add support for this.

### PR DESCRIPTION
**Pull request by DRK3**
Signed-off-by: Derek Trider <Derek.Trider@securekey.com>

closes #57

### CODE REVIEW

File: pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go [hunk 1]
```diff
@@ -94,28 +104,13 @@ func (c CouchDBEDVStore) Put(document models.EncryptedDocument) error {
...
-		MatchingEncryptedDocID: encryptedDocID,
-	}
-
-	documentBytes, err := json.Marshal(mapDocument)
-	if err != nil {
-		return err
-	}
-
-	return c.coreStore.Put(encryptedDocID+"_mapping_"+uuid.New().String(), documentBytes)
-}
```
Comment (DRK3) [inline pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go:110]: This was just moved down for consistency with the other unexported functions.

File: pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go [hunk 2]
```diff
@@ -137,7 +132,146 @@ func (c CouchDBEDVStore) Query(query *models.Query) ([]string, error) {
 	return c.filterDocsByQuery(idsOfDocsWithMatchingQueryIndexName, query)
 }
 
-func (c CouchDBEDVStore) findDocsMatchingQueryIndexName(queryIndexName string) (map[string]struct{}, error) {
+func (c *CouchDBEDVStore) checkNewDocumentForConflictingIndexWithAnyExistingDocs(
```
Comment (bstasyszyn) [inline pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go:135]: This name is too long. How about 'validateDoc'? Other funcs: validateAttributes, etc.

File: pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go [hunk 3]
```diff
@@ -67,12 +68,21 @@ func (c CouchDBEDVProvider) OpenStore(name string) (edvprovider.EDVStore, error)
...
 }
 
 // Put stores the given document.
 // A mapping document is also created and stored in order to allow for encrypted indices to work.
-func (c CouchDBEDVStore) Put(document models.EncryptedDocument) error {
+func (c *CouchDBEDVStore) Put(document models.EncryptedDocument) error {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	err := c.checkNewDocumentForConflictingIndexWithAnyExistingDocs(document)
```
Comment (bstasyszyn) [inline pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go:81]: This check isn't atomic, i.e. by the time the check finishes and no error returns another document with the same index could have been stored. What happens in this case, i.e. when you do a subsequent query using this index?

Comment (bstasyszyn) [reply pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go:81]: Sorry, I see that you actually do a write lock. I don't think it's a good idea to lock the entire store for a write of a single document. The store is supposed to handle many concurrent requests (if I'm not mistaken).

File: pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go (hunk 2 above)
Comment (DRK3) [reply pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go]: Done. Let me know if you think the new names are ok. I tried to make them as short as possible without ambiguity.

File: pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go (hunk 3 above)
Comment (DRK3) [reply pkg/edvprovider/couchdbedvprovider/couchdbedvprovider.go]: Per our discussion, I removed the lock and made a note regarding the impossibility of ensuring uniqueness due to the eventual consistency model of CouchDB. I will have to deal with the possibility of duplicate VCs in edge-service still - but I don't believe I can do anything about it here. This is just a best-effort attempt.

**Pull request merged**